integration/kotlinx-coroutines-androidx-lifecycle/src/Lifecycle.kt

```diff
@@ -0,0 +1,92 @@
+package kotlinx.coroutines.androidx.lifecycle
+
+import androidx.lifecycle.GenericLifecycleObserver
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.Lifecycle.State.DESTROYED
+import androidx.lifecycle.Lifecycle.State.INITIALIZED
+import androidx.lifecycle.LifecycleOwner
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Returns a [CoroutineScope] that uses [Dispatchers.Main] by default, and that is cancelled when
+ * the [Lifecycle] reaches [Lifecycle.State.DESTROYED] state.
+ *
+ * Note that this value is cached until the Lifecycle reaches the destroyed state.
+ */
+val Lifecycle.coroutineScope: CoroutineScope
+    get() = lifecycleCoroutineScopes[this] ?: job.let { job ->
```

Comment:
I don't believe this is thread safe.

I think the following scenario could occur:

- `Thread 1` reads `coroutineScope` and `this` is not in `lifecycleCoroutineScopes` collection, so it starts creating a new scope
- `Thread 2` reads `coroutineScope` and it **also** determines that `this` is not in `lifecycleCoroutineScopes` collection
- `Thread 1` finishes creating `CoroutineScope` and adds it to `lifecycleCoroutineScopes` collection
- `Thread 3` reads `coroutineScope` (and it retrieves it from `lifecycleCoroutineScopes` collection) and launches a Coroutine under the scope
- `Thread 2` finishes creating `CoroutineScope` and adds it to `lifecycleCoroutineScopes` collection
    - So we've lost reference to the `CoroutineScope` that `Thread 3` used to launch it's Coroutine

Reply:
@twyatt It is still thread-safe, because this is only for cache. This is not a problem if multiple job are created because of multi-thread access, as an attempt to remove them from the cache is performed anyway when the lifecycle is destroyed (which is independent from the cache that may be overridden).

Reply:
@twyatt In your scenario, all threads accessed a valid instance of a scope, and all of these instances will eventually be destroyed with the `LifecycleOwner`.